org/mozilla/jss/ssl/javax/JSSEngineReferenceImpl.java

```diff
@@ -0,0 +1,1170 @@
+        is_outbound_closed = true;
+    }
+
+    public String getHostname() {
+        return hostname;
+    }
+
+    public Runnable getDelegatedTask() {
+        debug("JSSEngine: getDelegatedTask()");
+
+        // We fake being a non-blocking SSLEngine. In particular, we never
+        // export tasks as delegated tasks (e.g., OCSP checking), so this
+        // method will always return null.
+
+        return null;
+    }
+
+    public SSLEngineResult.HandshakeStatus getHandshakeStatus() {
+        debug("JSSEngine: getHandshakeStatus()");
+        try {
+            updateHandshakeState();
+        } catch (SSLException se) {
+            // Do nothing -- updateHandshakeStatus() should throw the same
+            // exception again, so it is safe to discard it. We also can't
+            // just pass the SSLException to the caller of this method, as
+            // they aren't expecting it.
+        }
+
+        return handshake_state;
+    }
+
+    public SecurityStatusResult getStatus() {
+        if (ssl_fd == null) {
+            return null;
+        }
+
+        return SSL.SecurityStatus(ssl_fd);
+    }
+
+    /**
+     * Enable writing of encrypted TLS traffic to the specified port in a
+     * client-server relationship (mirroring the actual role of this
+     * SSLEngine) to enable debugging with Wireshark.
+     */
+    public void enableSafeDebugLogging(int port) {
+        debug_port = port;
+    }
+
+    private int computeSize(ByteBuffer[] buffers, int offset, int length) throws IllegalArgumentException {
+        debug("JSSEngine: computeSize()");
+        int result = 0;
+
+        if (buffers == null || buffers.length == 0) {
+            debug("JSSEngine.compueSize(): no buffers - result=" + result);
+            return result;
+        }
+
+        // Semantics of arguments:
+        //
+        // - buffers is where we're reading/writing application data.
+        // - offset is the index of the first buffer we read/write to.
+        // - length is the total number of buffers we read/write to.
+        //
+        // We use a relative index and an absolute index to handle these
+        // constraints.
+        for (int rel_index = 0; rel_index < length; rel_index++) {
+            int index = offset + rel_index;
+            if (index >= buffers.length) {
+                throw new IllegalArgumentException("offset (" + offset + ") or length (" + length + ") exceeds contract based on number of buffers (" + buffers.length + ")");
+            }
+
+            if (rel_index == 0 && buffers[index] == null) {
+                // If our first buffer is null, assume the rest are and skip
+                // everything else. This commonly happens when null is passed
+                // as the src parameter to wrap or when null is passed as the
+                // dst parameter to unwrap.
+                debug("JSSEngine.compueSize(): null first buffer - result=" + result);
+                return result;
+            }
+
```

Comment:
This is a touch subtle. I'm happy to add a better comment if you can think of one.

At the lines at the start of this function, I see:

```java
        if (buffers == null || buffers.length == 0) {
            debug("JSSEngine.compueSize(): no buffers - result=" + result);
            return result;
        }
```

In this case, if buffers itself is `null`, the size of the buffers itself is 0. There are no buffers, therefore the length is 0.

----

So then we get here:

```java
            if (rel_index == 0 && buffers[index] == null) {
                // If our first buffer is null, assume the rest are and skip
                // everything else. This commonly happens when null is passed
                // as the src parameter to wrap or when null is passed as the
                // dst parameter to unwrap.
                debug("JSSEngine.compueSize(): null first buffer - result=" + result);
                return result;
            }
```

What happens is that there's three forms of `wrap` and `unwrap` calls:

 - `wrap(ByteBuffer src, ByteBuffer dst)` -- single `src`/`dst` args
 - `wrap(ByteBuffer[] srcs, ByteBuffer dst)` -- array of `src`s, single `dst`
 - `wrap(ByteBuffer[] srcs, int offset, int length, ByteBuffer dst)` -- array of `src`s, single `dst`, and indices for accessing `srcs` at. 

In particular, `SSLEngine` converts the first three forms into the latter form. So if you call:

```java
ByteBuffer src = null;
ByteBuffer dst = null;
engine.wrap(src, dst);
```

It gets translated:

```
public SSLEngineResult wrap(ByteBuffer src, ByteBuffer dst) {
    return this.wrap(new ByteBuffer[] { src }, 0, 1, dst);
}
```

And so the `null` is passed through, giving us an array of buffers with length 1, but whose size is still 0 (because the first element is a `null` `ByteBuffer`). So we return 0.

----

Finally, we get to this check:

```java
            if (buffers[index] == null) {
                throw new IllegalArgumentException("Buffer at index " + index + " is null.");
            }
```

This check is equivalent to:

```java
            if (rel_index > 0 && buffers[index] == null) {
                throw new IllegalArgumentException("Buffer at index " + index + " is null.");
            }
```

(because we would've returned above if `rel_index == 0` and `buffers[index] == null`. So it can only be true when `rel_index > 0`. 

In this case, we've gotten past the first element and _then_ found a `null` `ByteBuffer` at some higher index.

We thus had two choices:

 1) Strictly counted the actual size of all `ByteBuffers` (skipping `null` where present),
 2) Treat this as an error and only accept `null` in the first position due to the above translation.

Since the way the `SunJSSE` `SSLEngine` implementation works is to not accept `null` at all _anywhere_, we're already more lenient and graceful in handling these cases. And we went with route #2, where this is behaving as expected, throwing an error.

So I'm inclined to leave this as-is, unless you have a strong preference towards 1.